### ISSUE TITLE
1.3.0 - Only root can write to OSX volumes / Can't change permissions within

### ISSUE THREAD

**Issue by lynxnathan** [boot2docker/boot2docker#581]
Currently only root inside the container is able to modify mounted volumes. Also permissions changed from within have no effect. I'm unsure as to whether `cannot be managed dynamically` on the blog post refers to any of this.

I would suggest updating the readme to let users know about the existence of native support for volumes and also its limitations (especially if this is one of them).

**Comment by brunoric** [comment]
Same problem here. =(

**Comment by aladagemre** [comment]
I have the same problem with boot2docker: Docker v1.4.1, fig 1.0.1
It's fine on Linux: Docker v1.4.0, fig 1.0.1

I can't mount rabbitmq or mongodb just for persistence. 
Any news on this?

**Comment by fcvarela** [comment]
Not sure how helpful this is but i always create a user w/ id 1000 in the
container and run the processes from that. Works perfectly fine. I also ADD
the data needed and them add a VOLUME to the same location. You can choose
to -v vol:vol in production and not to in dev, thereby using the same image
for both envs.

Hope this helps

On Tue, Jan 6, 2015 at 12:00 PM, Ahmet Emre Aladağ <notifications@github.com

**Comment by fcvarela** [comment]
Actually the other way around, you can docker run -v stuff:stuff in dev and
not to in prod :) Sorry about that.

**Comment by hgl** [comment]
I think this should be fixed from the docker end. Related issue: https://github.com/docker/docker/issues/7198

Every time a host directory is mounted, you should be able to specify uid/gid of the mounted directory. That way, boot2docker can keep using whatever uid & gid for vboxsf, and still allow different users in different containers to write to it.

A lot of people affects by this, not sure why that issue is still not fixed, probably not very easy to implement?

**Comment by sjtuadamyang** [comment]
@vmaatta , so it's not possible to set up boot2docker + postgres with /var/lib/postgresql/data mounted from host machine since postgres needs to create hardlink in /var/lib/postgresql/data which is not supported by vboxsf??

**Comment by vmaatta** [comment]
@sjtuadamyang Pretty much, yes. You should probably go with a data volume container for Postgres. In the data volume container make `/var/lib/postgresql/data` just a normal volume. Additionally add some other path such as `/mnt/hostdata` as a host volume to some location on your host.

Now, as mentioned you still have to get data into and out of the database but it's not as bad as it sounds in Postgres' case. Just like you can open `psql` in a container, you can `pg_dump` data in a temporary container into/out the extra host mounted folder:

`docker run -ti --volumes-from postgres-data --link postgres:postgres --rm postgres sh -c 'exec pg_dump -h "$POSTGRES_PORT_5432_TCP_ADDR" -p "$POSTGRES_PORT_5432_TCP_PORT" -U postgres -Fd -f /mnt/hostdata/backup/some_database_local_20150120_bak <some_database>'`

That assumes there's a data volume container `postgres-data` and a running Postgres database in container `postgres`. All of these containers are `postgres:latest`.

**Comment by sjtuadamyang** [comment]
@vmaatta , thanks for you explanation. Correct me if I understand it wrong, the reason I need a data-only container here is that I can keep the data-only container alive, and just run the postgres container one-off to dump the data into host directory. Theoretically, I can do the same thing by just get into the postgres container when it's serving and dump the database by issue the pg_dump command inside the postgres container, right?

**Comment by vmaatta** [comment]
@sjtuadamyang Yes, you can make it work without a data volume container; you can just change the `--volumes-from` to your running postgres and otherwise modify as needed. Generally I would not "get into", i.e. SSH (or nsenter etc.) into a container.

You can read up about data volume containers and why maybe use them in the [documentation](https://docs.docker.com/userguide/dockervolumes/) and only you can really know whether to use them or not.

**Comment by neciu** [comment]
Same here lads.

**Comment by pkerpedjiev** [comment]
Pardon my ignorance, by why doesn't the mounted folder within the container have group write permissions? Other users could then be added to the 'staff' group and everything would be much easier.

**Comment by Vrakfall** [comment]
Thanks @bradgessler for that workaround, I really like it. I'll use a groupadd instead as it seems to be a less violent solution to me.

The `--vbox-share=[]` option of `boot2docker` seems to be another good solution to me if you know the uid in advance but I still prefer a groupadd to this.

I see no other solution than the workaround because, for me, it all comes from Virtualbox, as @paolomainardi said. It would be better Virtualbox supports POSIX permissions "on the go". Sad that remounting doesn't actually work. :'(

I checked a bit, but not really much, and I didn't find this "usermod/groupadd" workaround in the boot2docker documentation. Shouldn't it be there for less users to lose time while trying to figure out why mysql (e.g.) can't write on its on database ?

**Comment by paolomainardi** [comment]
@Vrakfall switch to NFS to fix those issues.

**Comment by Vrakfall** [comment]
Mmmh, what about performances ? Isn't vboxfs better for hard hdd usages ?

**Comment by paolomainardi** [comment]
@Vrakfall not really: http://mitchellh.com/comparing-filesystem-performance-in-virtual-machines nfs/samba is ever a better option.

**Comment by Vrakfall** [comment]
Wow, I wasn't expecting such a difference !
Thanks a lot ! You made me change my mind for the NFS solution here ! =D (Would be hard for Windows users btw).

**Comment by Vrakfall** [comment]
Arh ! Any clue or howto about mounting nfs from host to a container ? I find myself stuck because my container can't ping my server.

Thanks in advance !

**Comment by darwingr** [comment]
Use Sven's 'Folder Sharing' instructions in the readme but instead of cifs put :+1: 
`nfs://192.168.59.103/data`
Then just copy or clone your files into that network share I guess.

**Comment by Vrakfall** [comment]
Hosting folders with samba to read them with nfs... Meh, that sounds really weird to me. Ok, that would work but... I'd really much prefer using nsf in this case. Any clue ? @paolomainardi ?

**Comment by paolomainardi** [comment]
@Vrakfall if you are using osx:

1) Edit `/etc/exports` as follows:
```
/Users -mapall=[youruser]:staff [boot2docker-vm-ip]
```

Where [youruser] is your localuser as reported by `whoami` and boot2docker-vm is the ip as reported by `boot2docker ip`

2) Restart nfsd services:

```
sudo nfsd stop && sudo nfsd start
```

3) Create a script on boot2docker vm on `/var/lib/boot2docker/bootlocal.sh`: 

```
docker@boot2docker:~$ cat /var/lib/boot2docker/bootlocal.sh
#|/bin/bash
sudo umount /Users
sudo /usr/local/etc/init.d/nfs-client start
sudo mount 192.168.59.3:/Users /Users -o rw,async,noatime,rsize=32768,wsize=32768,proto=tcp
```

You're done.

**Comment by paolomainardi** [comment]
And in the meanwhile i'm working on improving NFS/CIFS (samba) shares using FSCACHE kernel implementation, for boot2docker.
If you are curious/interested follow this thread: https://github.com/boot2docker/boot2docker/pull/699

**Comment by Vrakfall** [comment]
Thanks for the help. Yes, I'm working on osx at work.
It feels I should copy it as root to `/var/lib/boot2docker/bootlocal.sh` but I can't find the root password anywhere. The docker user can't write to that directory or am I doing something wrong ?

I tried this:
```
$boot2docker ssh
$cp /Users/osxUser/path/bootlocal.sh /var/lib/boot2docker/bootlocal.sh
cp: overwrite '/var/lib/boot2docker/bootlocal.sh'? y
cp: can't create '/var/lib/boot2docker/bootlocal.sh': Permission denied
```

I'm forced to using command line because my goal is to create an automated script.

**Comment by paolomainardi** [comment]
Just try "sudo su"

Sorry for typos, sent by mobile.
Il 17/feb/2015 10:11 "Jérémy Lecocq" <notifications@github.com> ha scritto:

**Comment by Vrakfall** [comment]
What was I thinking. xD

Oh and thank you for #699. I'm indeed curious AND interested in any better way of sharing host's files to containers.

I'm still stuck trying to mount the nfs share:
```
docker@boot2docker:~$ sudo sh /var/lib/boot2docker/bootlocal.sh
mount: RPC: Remote system error - Connection refused
mount: mounting 192.168.59.3:/Users on /Users failed: Bad file descriptor
```
Weird because here is my `/etc/exports` file:
```
/Users -mapall=stagiaire:staff 192.168.59.103(rw,non_root_squash,no_subtree_check)
```
I'm used to Linux for a long time but I'm new to osx so I might be missing some obvious points.

**Comment by paolomainardi** [comment]
Try to start nfs-rpc services:
/usr/local/etc/init.d/nfs-client start

Sorry for typos, sent by mobile.
Il 17/feb/2015 10:45 "Jérémy Lecocq" <notifications@github.com> ha scritto:

**Comment by Vrakfall** [comment]
It was done by the `.sh` script. Sorry, I've removed the obvious lines from the output juste to make it clearer.
Here the original one:
```
docker@boot2docker:~$ sudo sh /var/lib/boot2docker/bootlocal.sh
umount: can't umount /Users: Invalid argument #Because I tried it multiple times
/usr/local/sbin/rpcbind is already running
808
/usr/local/sbin/rpc.statd is already running
810
Starting nfs client utilities.
mount: 192.168.59.3:/Users failed, reason given by server: Permission denied
mount: mounting 192.168.59.3:/Users on /Users failed: Bad file descriptor
```

**Comment by paolomainardi** [comment]
I see a "permission denied" on osx side.
Are you sure that boot2docker IP on export is correct ? Have you tried to
restart nfsd on osx ?

Sorry for typos, sent by mobile.
Il 17/feb/2015 10:53 "Jérémy Lecocq" <notifications@github.com> ha scritto:

**Comment by Vrakfall** [comment]
Yup for both statements.
```
$ boot2docker ip
192.168.59.103
```

**Comment by paolomainardi** [comment]
Try to remove the options you specified () on /etc/exports and restart the
server.

Sorry for typos, sent by mobile.
Il 17/feb/2015 11:03 "Jérémy Lecocq" <notifications@github.com> ha scritto:

**Comment by Vrakfall** [comment]
That did it. :3 It was kinda my fault. I've removed the async option from `/var/lib/boot2docker/bootlocal.sh` and it worked as soon as I put it back (and after removing options from `/etc/exports`).
I'm not used to nfs servers. :s
I was thinking about moving the discussion to the emails or by skype or any IM way because we kinda were spamming everyone. =/ Sorry for that. But now it's working so it's too late.

Thank you a lot again !

**Comment by ludofleury** [comment]
@mnapoli for your use case, I ended with a solution mixing volume container & custom path for the Symfony/php app cache/logs.

I've got a volume container for /var/cache/my-app & another one for /var/log/my-app.
Pretty neat.

**Comment by mikehaertl** [comment]
> I checked a bit, but not really much, and I didn't find this "usermod/groupadd" workaround in the boot2docker documentation.

+1

The `usermod -u 1000 www-data` (or even `usermod -u 1000 -g 1000 www-data`) fixes it for us. This should really be added as a simple workaround to the documentation.

**Comment by cordoval** [comment]
just in case i think @ludofleury meant this http://www.craftitonline.com/2015/02/docker-symfony-development-box-with-data-containers/ perhaps it can be useful to some that ran into this.

**Comment by mnapoli** [comment]
@ludofleury thank you! I need to give another try at docker… That whole boot2docker with virtualbox + problems with write permissions was really a cold shower, lost so much time on this and then gave up… I need to find another week end to spare :p

**Comment by Vrakfall** [comment]
@mnapoli If you find yourself stuck with permissions, try mounting your files on the boot2docker vm with nfs rather than vbox shares. ;D It's pretty neat and supports uid and gid permissions.

**Comment by danros** [comment]
I simply put this in my /var/lib/boot2docker/profile 

umount /Users
mount -t vboxsf -o uid=33,gid=33 Users /Users

This makes sure /Users is readable and writable by the www-data user.

**Comment by fieder** [comment]
Same here. +1 for a fix! 

I'm running iOS. I'm trying to implemented data-only container pattern. First container with the volume gets created OK. 

Then my second container runs RabbitMQ. The idea is to forward RabbitMQ's /logs (logfiles) and /mnesia (persisted data) to /data directory which is mounted to the RabbitMQ container from the data-only container via

docker run  -d --volumes-from=my-data-only-container  --name my-rabbitmq -P myrepo/rabbitmq

As a result I get attached to my-rabbitmq the /data directory but the start fails because of permissions over that folder. Rabbitmq can't write the logs, pid files, etc. If I access my-rabbitmq container with bash, I can see the permissions totally wrong
ls -la /data gives me 
1000 , staff  
which of course does not match any user/group in my container. 

Workaround 
RUN usermod -u 1000 rabbitmq seems to fix it but then if I want to create another container, say mysql pointing to the same data-only container, I will have to do RUN usermod -u 1000 mysql too and the same for any other service (mongo, postgress, you name it!) in which case I don't think this workaround will work for more than 1 service container pointing to the same data container.

**Comment by ludofleury** [comment]
@cordoval @mnapoli yes, exactly I gave a less detailed example there: http://testonsteroid.tumblr.com/post/111468785847/docker-symfony-mac-filesystem-permissions

Yet, I abandoned docker for the moment. I was trying to setup a full workflow between local env to production env with it, and it's not ready for that now. I'm back to the classic stack with separated VM / CI / prod env, mixing vagrant/berkshelf and puppet.

**Comment by cordoval** [comment]
sorry to hear you abandoned it

**Comment by sskorc** [comment]
Thanks @danros for sharing [your workaround](https://github.com/boot2docker/boot2docker/issues/581#issuecomment-78030456). It works perfectly!

**Comment by grimabe** [comment]
@danros workaround is working fine, but is there any fix coming ?

**Comment by danros** [comment]
I am not involved with the boot2docker project, but I suspect it won't be fixed, when the solution is a couple of lines in a file designed for exactly this kind of customisation. It should probably be documented though.

**Comment by c0nstantx** [comment]
:+1: @danros This workaround did the trick and saved me from insanity! Thank you.

**Comment by exlee** [comment]
I think I got a fix for that.

Proof of concept code (it's ugly, but that's due my Shellscript inaptitude - and I wanted to use shellscript so that other could use that too):
https://gist.github.com/exlee/4b35915bbb96aa9ac8b8

What I'm doing here is using separate Mountfile to specify how directories should be linked. Then, based on this, I'm disabling machine, adding shares using auto-generated ids and mounting them inside while using given uid/gid (which, not to say has to be exactly the same as the one used inside Docker).

I've explored this issue for quite a while, and I think that is the proper way to approach it. I would prepare patch myself, but don't speak Go :disappointed:

**Comment by perrierism** [comment]
> but I suspect it won't be fixed, when the solution is a couple of lines in a file designed for exactly this kind of customization

The workaround @danros suggests only works for one user. If you have one container that needs access for the `www-data` user it works ok, but once you need to simultaneously run another container which needs a volume mounted from anywhere your filesystem with access for `another-user` it won't work.

**Comment by ljsommer** [comment]
+1 this is a pain. Trying to dev postgresql on osx using vagrant mounting data volume from a data container into my process container with no luck.
Why is Docker so painful  =(

**Comment by alasarr** [comment]
+1. Also with postgresql. That would be really nice to fix this issue.

**Comment by rtancman** [comment]
+1 RUN usermod -u 1000 mysql

@bradgessler thx ;)

**Comment by dsummersl** [comment]
+1 for the official mongo image - I haven't been able to get the usermod fix to work as that image uses root...arg!

**Comment by pdonorio** [comment]
+1 for a real fix, this is a very serious issue for mac developers!

I really don't like "usermod -u 1000 postgres" (and I did not happen to make it work, also combined with groupmod) since i use a lot docker compose and official images, which i do not want to modify. I prefer to add volumes and environment variables for customization, i feel like it's such a better practice.

I failed also when trying the mount fix: it fails in various way, for example i get at bootstrap:
```
FATAL:  could not open file "pg_xlog/000000010000000000000001": No such file or directory
```

How can we all help?

**Comment by mnapoli** [comment]
@danros it seems for me that boot2docker doesn't use the `/var/lib/boot2docker/profile` file. However if I run those commands manually *inside* the boot2docker VM (`boot2docker ssh`) it works:

```
umount /Users
mount -t vboxsf -o uid=33,gid=33 Users /Users
```

Anyone knows where to store these commands in a file so that boot2docker runs this every time?

**Comment by tomdavidson** [comment]
@mnapoli you can make your own [custom boot2docker.iso](https://github.com/boot2docker/boot2docker/blob/master/doc/BUILD.md) fairly easy. boot2docker uses tinycore, so you can make persistent changes the [tinycore way](http://www.brianlinkletter.com/persistent-configuration-changes-in-tinycore-linux/).

**Comment by nachimehta** [comment]
I was getting the same problem before when trying to share volumes in a mysql container. I added RUN usermod -u 1000 mysql to the Dockerfile.

Now I'm getting a unix port bind error:

    [ERROR] Can't start server : Bind on unix socket: Permission denied
    [ERROR] Do you already have another mysqld server running on socket: /var/run/mysqld/mysqld.sock ?
    [ERROR] Aborting

Anyone else encounter this?

**Comment by perrierism** [comment]
@nachimehta 

Yes, everyone attempting to mount volumes for mysql containers encounters this.

**Comment by mnapoli** [comment]
After using the hack mentioned here, I realized that vboxfs performances are just too terrible. It's unusable for bigger applications.

I am now using NFS mounts instead. The whole thing is described here: http://syskall.com/using-boot2docker-using-nfs-instead-of-vboxsf/ (the instructions in this thread were incomplete for me).

No permission issues (for now) and a little faster I/O.

**Comment by paolomainardi** [comment]
Hi guys!

If you want a better and native NFS implementation for boot2docker, please
test this MR: https://github.com/boot2docker/boot2docker/pull/699

Sorry for typos, sent by mobile.
Il 07/giu/2015 20:12 "Matthieu Napoli" <notifications@github.com> ha
scritto:

**Comment by mnapoli** [comment]
Yet another update people: I've started using [Dinghy](https://github.com/codekitchen/dinghy) instead of Boot2docker and it rocks!

It's amazingly fast (it uses NFS, but I don't know why it's even faster than boot2docker with NFS on my machine) and no permission issues.

**Comment by pperez** [comment]
@mnapoli Can you use images that do chmod/chown?

**Comment by mnapoli** [comment]
@pperez I just tried `docker-compose run mycontainer chmod 777 /app/tmp/logdump.sql` and the file went from `-rw-rw-r--` to `-rwxrwxrwx` (on OS X) so I guess the answer is yes :)

**Comment by jakedeichert** [comment]
I just spent about 8 hours today at work trying to get mysql to use a volume on osx through boot2docker... then finally it worked using a couple of different RUN commands in a Dockerfile. As some people have mentioned, `usermod -u 1000` works but I also had to change a group and an owner as you'll see below.

Dockerfile
~~~
FROM mysql
RUN usermod -u 1000 mysql
RUN usermod -G staff mysql
RUN chown mysql /var/run/mysqld
EXPOSE 3306
~~~

Then build and run it...
~~~
docker build -t custom-mysql .
docker run -d --name db -p 3306:3306 -v /Users/<name>/path:/var/lib/mysql custom-mysql
~~~

After 8 hours, I saw [this dockerfile](https://registry.hub.docker.com/u/dgraziotin/mysql/dockerfile/) and realized I never tried `usermod -G` (after a million other attempted solutions). 8 hours and all I needed was to change the group :-1:

**Comment by tompson** [comment]
@jakedeichert thanks for your post - I just extended the official wordpress image with

    FROM wordpress
    RUN usermod -u 1000 www-data
    RUN usermod -G staff www-data

to allow me mounting the wordpress folder from OS X to /var/www/html and installing plugins

**Comment by luxifer** [comment]
For me it's still just a workaround to have to run usermod inside the dockerfile. What if you want to share the image? push it to the docker hub? Or use it anywhere else but your laptop? It should only be a platform problem, not a user problem. I'm still looking for real solutions (like NFS) or maybe another VM like https://github.com/codekitchen/dinghy. For the NFS part, take a look at http://blog.blackfire.io/how-we-use-docker.html

**Comment by pdonorio** [comment]
I totally agree with @luxifer: IMHO if we need a workaround for any image to work properly on the laptop, then we are back to where we lived before Docker.

Also another VM would become a real option for my self only when integrated in docker machine, which is the best way at the momento to have a development environment on the laptop which you run with the same commands on the most famous cloud provisioning, including my own internal openstack cloud. And for dinghy we have to wait: https://github.com/codekitchen/dinghy/issues/16

Looks like the best bet is NFS, will run some tests.

**Comment by jakedeichert** [comment]
Yeah it's definitely a temporary workaround. I'll be waiting for an actual solution so this doesn't have to be done per-container.

**Comment by mnapoli** [comment]
@pdonorio is there anything preventing us to use docker-machine instead of boot2docker/dinghy? I just read about it and it looks like a replacement rather than an addition.

**Comment by pdonorio** [comment]
@mnapoli docker machine uses virtualbox + boot2docker as local host
e.g. see https://docs.docker.com/machine/ and search "Oracle VirtualBox".

I am using machine/virtualbox and facing the same problems described in this issue, and dinghy does not support machine yet.

**Comment by yoanisgil** [comment]
I run into this same issue and since I did not want to bang my head against this for too long I went ahead with a *ugly not recommended* hack:

- Get yourself a copy of /etc/passwd running inside your container 
- Modify the file you just copied so that id's match whatever user you want to map
- Map the modified with docker run -v

In my case it was all about mapping user with id 1000 to www-data within a container so the modified/hacked line looks like this:

     www-data:x:1000:1000:www-data:/var/www:/usr/sbin/nologin

after that I ran:
     
     docker-run -v $(pwd)/passwd:/etc/passwd myimage

and it did the work. Not the best solution ever but it let me move on onto more important stuff ;)

**Comment by perrierism** [comment]
@mnapoli It looks like Machine still uses boot2docker for local OSX and Win environments.

**Comment by timbrandin** [comment]
@tompson's https://github.com/boot2docker/boot2docker/issues/581#issuecomment-114804894 solved it for me: https://github.com/timbrandin/docker-drupal-kitematic/blob/master/Dockerfile#L27

**Comment by saada** [comment]
@timbrandin , you're a life saver! :bow:

**Comment by luxifer** [comment]
I tried vagrant-coreos and it works like a charm. With NFS.

Florent Viel
Le 11 juil. 2015 11:59 PM, "Mahmoud Saada" <notifications@github.com> a
écrit :

**Comment by marklocker** [comment]
@nachimehta I'm experiencing the same as you. After adding `RUN usermod -u 1000 mysql` to my MySQL Dockerfile, the container now writes to my host OS but I get the following error:

    [ERROR] Can't start server : Bind on unix socket: Permission denied
    [ERROR] Do you already have another mysqld server running on socket: /var/run/mysqld/mysqld.sock ?
    [ERROR] Aborting

Did you find a solution to this?

**Comment by nachimehta** [comment]
@marklocker ,

I went with the following:
    `ssh boot2docker`
    `sudo mount -t vboxsf -o uid=999,gid=50 Users /Users`

it's not pretty, and you have to do it each time you re-init your docker container, but it works.

**Comment by mnapoli** [comment]
@marklocker Try https://github.com/codekitchen/dinghy permissions just work correctly, and it uses NFS so it's much faster.

**Comment by marklocker** [comment]
@mnapoli Funnily enough I tried Dinghy based on some messages I'd seen you post, but experienced the same issue. Apparently it used to be fine but a change was made which introduced the same issue that boot2docker has, and it's not one they want to go back on because it sorts out some other problems.

**Comment by termosa** [comment]
@mnapoli @marklocker same for me

**Comment by mnapoli** [comment]
@termosa @marklocker oh weird, I'm using it as we speak ([updated and upgraded](https://github.com/codekitchen/dinghy#upgrading) as of 2 days ago) and I have no issues. Hopefully it stays that way, I've invested too much in Docker now to go back :'(

**Comment by saada** [comment]
@mnapoli what version of OS X are you on?

**Comment by mnapoli** [comment]
@saada Yosemite.

**Comment by termosa** [comment]
@saada as I've got from the discussion above it's not about OS-X, but it's about the way of mounting in boot2docker.

**Comment by paolomainardi** [comment]
Just a quick update from my experiments, some weeks ago has been merged my PR about including fscache modules on boot2docker, starting from 1.7.0, this means that NFS+FSCache can achieve almost native FS performance.

MR: https://github.com/boot2docker/boot2docker/pull/699

How to test it:

1) Configure NFS on OSX side.

Edit `/etc/exports` as follows:

```
/Users -mapall=YOURUSER:staff BOOT2DOCKER_IP
```

Replacing `YOURUSER` with the result of `whoami` and BOOT2DOCKER_IP with the result of `boot2docker ip`

Now restart NFS service: `sudo nfsd stop && sudo nfsd start`

2) Run cachefilesd container:

```
boot2docker ssh "sudo modprobe cachefiles"
docker run --privileged -d -v /mnt/sda1/var/cache/fscache:/var/cache/fscache sparkfabrik/docker-cachefilesd
```

You should see something like this:

```
% docker ps                                                                                                                                                                              
CONTAINER ID        IMAGE                               COMMAND                CREATED              STATUS              PORTS                    NAMES
5cc4416c9f58        cachefilesd:latest                  "/bin/sh -c '/sbin/c   13 minutes ago       Up 13 minutes                                lonely_goldstine
```

3) Unmount vboxfs and mount NFS with `fsc` attribute on:

```
boot2docker ssh "sudo umount /Users"
boot2docker ssh "sudo /usr/local/etc/init.d/nfs-client start"
boot2docker ssh "sudo mount.nfs 192.168.59.3:/Users /Users -o fsc,rw,async,noatime,rsize=32768,wsize=32768,proto=tcp"
```

It's done, check if it works correctly, you should see something like this:

```
% boot2docker ssh 'sudo cat /proc/fs/nfsfs/volumes'                                                                                                                             
NV SERVER   PORT DEV     FSID              FSC
v3 c0a83b03  801 0:72    1000003:0         yes

% boot2docker ssh 'sudo du -chs /mnt/sda1/var/cache/fscache'
16.3M   /mnt/sda1/var/cache/fscache
16.3M   total
```

**Comment by mnapoli** [comment]
@paolomainardi interesting, do you have any numbers on how faster the filesystem is? I'm just wondering if the additional steps are worth it since NFS "feels" good enough (compared to vboxfs which is just a show-stopper).

**Comment by paolomainardi** [comment]
@mnapoli this article has been very helpful: http://chase-seibert.github.io/blog/2014/03/09/vagrant-cachefilesd.html

**Comment by YGuillemin** [comment]
@paolomainardi The way to use nfs instead of vboxfs allow to avoid all problems about permissions ?

If i mount a volume from my OSX like 
     `/private/var/www -> /var/www` OSX to boot2docker 
And then i want to launch a container and attach that volume like
   `docker run -ti --rm -v /var/www:/var/www php-fpm` (attach on php-fpm or nginx container for example)
The container php-fpm has permissions problem to clear:cache in symfony, but no with nfs ?
i've tried to change iud when i mount the boot2docker volume : 
   `sudo mount -t vboxsf -o uid=1000,gid=33 Users /Users`
And i've changed the iud of www-data in my Dockerfile : 
   `RUN usermod -u 1000 www-data`
But i can't clear the cache anymore if i launch a page from the server, at this url for example: 
  `BOOT_2_DOCKER_IP/web/app_dev.php`
Generated files haven't the same permissions ? 

You said that nfs is faster than vbox, but does it allow any container which share volume to write within too ?

Any solution works for me, but i've not tried nfs yet....

btw, sorry for my bad english...

**Comment by paolomainardi** [comment]
yes @YGuillemin it permits to avoid almost any permission problem.

**Comment by YGuillemin** [comment]
Thanks for your help.

Indeed, it's very slow even with nfs...i'll go back later (2, 3 months) in order to take a look to docker, when it'll come up ;)

**Comment by nguyentam2211** [comment]
Hi there,

For ones who are using Kitematic, just do this:
1. Open your Kitematic application
2. Open Settings tab from your container
3. Add this value to Environment Variables:
- Key: "USERMOD"
- Value: "1000 www-data" (NOT 1000:www-data)

Hope that can help you about this permission issue!

**Comment by frank-dspeed** [comment]
how about starting docker and boot2docker as root in OSX i am not a OSX user but got locked into that by a costumer i think when boot2docker and docker itself would get used with osx sudo then all would work as expected can some one tell me who is responsible for boot2docker at present is it tianon?

**Comment by mnapoli** [comment]
Anyone tried the new Docker Toolbox on Mac? Did it fix the permission issues or is it still the same?

**Comment by nachimehta** [comment]
@mnapoli , I did. It just uses docker-machine which is still built using the boot2docker iso. Problem still exists.

**Comment by krasilich** [comment]
@mnapoli Hi. Any update of Docker will not fix this issue (at least while docker demon is not able to run on OS X directly), because issue comes from different level. Shared folders implementation of your VM provider is responsible.

**Comment by frank-dspeed** [comment]
i dont agree with that the boot2docker method is simply bad maybe a more good way would be to make a new boot2docker iso based on a stripped down kernel most current release would do the trick more well i am working at present on such a image to replace coreos and all this other solutions i think a basic kernel img with all needed kernel moduls for docker will be the most best fit then implament virtualbox guest addtions support as also fusion (vmware) support as option will be the successor over the boot2docker way and will lead into all functions work like expected.

If the guest additions dont do fix that then i will simply add casper fs for this and that method will work 100%

**Comment by saada** [comment]
@frank-dspeed, do you have a repo created somewhere for that?

**Comment by snowwolf007cn** [comment]
@krasilich Any way to fix this? If I can't mount the directory, any change to db will lost.

**Comment by frank-dspeed** [comment]
@saada sure here you got https://developer.ubuntu.com/en/snappy/

**Comment by Kephryn** [comment]
[paolomainardi's solution works great] (https://github.com/boot2docker/boot2docker/issues/581#issuecomment-127972657). 

If you are using Kitematic you can basically replace `boot2docker ssh` with `docker-machine ssh default`

Last thing to note is the change of IP address when doing the `mount.nfs` use `192.168.99.1`

@paolomainardi Thank you very much, I am very pleased everything is working and quick as well.

**Comment by jayfk** [comment]
This issue should be a top priority for the next feature release, or at least a documented `official` way to solve it on your own.

That's basically the first thing every new user runs into when building something with a database and that's a no go.

Sorry for the harsh response, but I had 3 mails and two tickets related to this today alone. I know you are working hard on the whole eco system, but docker is on its way to mass adoption by users that assume that basic things work and having a *persistent* database is one of these things.

**Comment by saada** [comment]
It's been almost one entire year!
:sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob:

**Comment by md5** [comment]
I personally think a good way to solve this issue would be to introduce a volume driver that's able to provision new VMDK or VDI disks under VirtualBox and to steer users away from using host directories for their data mount.

I started writing such a volume driver over at [`appropriate/docker-volumes-virtualbox`](https://github.com/appropriate/docker-volumes-virtualbox), but it ended up being really tedious to interact with VirtualBox's SOAP-based `vboxwebsrv`. There are also questions of how the lifecycle of `vboxwebsrv` would be managed in a `docker-machine`/`boot2docker` setup without having the user have to start and manage the web service themselves. Another approach I'm considering is creating a storage driver for the REX-Ray volume driver instead of creating a standalone volume driver for VirtualBox. See here: https://github.com/emccode/rexray/issues/104

**Comment by David-Crty** [comment]
My php-fpm container can't write in my app cache folder wich is a big probleme !

I can't use docker on my mac...

**Comment by saada** [comment]
@David-Crty, I have a mac too and here's what I have in my fpm Dockerfile I added this:
```
RUN usermod -u 1000 www-data
RUN usermod -G staff www-data
RUN chown -Rf www-data.www-data /usr/share/nginx/html/
```
My fpm runs as www-data user. My boot2docker VM user (host VM) runs as uid 1000.
This way I guarantee that files saved are owned by the same uid as boot2docker user and everything just works.

This is a terrible workaround because it allows inconsistencies between environments since not all host machine will run as user with uid 1000. It breaks the entire point of what docker promises which is "build once, run anywhere". It's all a big lie basically until they fix the one year old bug.

Our startup is struggling to go to production because of this bug. AWS user uids are not the same as our DEV machine uids which breaks permissions significantly. Our dockerfile does work the same everywhere. Anyone else run into this?

**Comment by ludofleury** [comment]
Use Docker-compose, extends the base config with a dev config, in the Dev
config use a data container.

**Comment by vielmetti** [comment]
Ran into this last night trying to get https://github.com/vpicavet/docker-pggis up and running. Thanks to @thaJeztah for identifying the issue. Fortunately there's a native Postgres.app available for OSX so I'm going to punt on this issue and move on.

**Comment by dmitrijev** [comment]
On the same boat with you guys, docker container still can NOT write to shared volume on Mac (10.11)

**Comment by nhooey** [comment]
Happy Birthday Bug 581!

![Birthday Cupcake](http://www.realfoodgirlunmodified.com/wp-content/uploads/2013/01/birthday-cupcake.jpg)

**Comment by saada** [comment]
:birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada:

**Comment by bmurphy1976** [comment]
I know everybody is as disappointed as I am that this issue does not yet have a resolution.  That said, could we please stop posting unproductive comments?  I am subscribed to this issue so that I  can track its progress not so that I can receive emoji spam in my email.  Thanks!

**Comment by perrierism** [comment]
:birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada: :birthday: :gift: :balloon: :tada:

**Comment by SobiborTreblinka** [comment]
+1

**Comment by krasilich** [comment]
Finally got the workaround that kinda works for me.

- Get [docker-machine-nfs](https://github.com/adlogix/docker-machine-nfs) script
- Run
~~~
$ docker-machine create -d virtualbox dev-nfs
~~~
- And then
~~~
$ docker-machine-nfs dev-nfs
~~~
This returns me
~~~
[INFO] Verify NFS mount ... 			FAIL
~~~
but share is created anyway.

- Open `/etc/exports` and replace `-mapall=$uid:$gid` with `-maproot=0`
- Run
~~~
$ sudo nfsd restart
~~~

Hopefully last two points will not be needed when (https://github.com/adlogix/docker-machine-nfs/issues/17) is fixed

- And finally
~~~
$ eval "$(docker-machine env nfs)"
$ docker run -d -P -e MYSQL_ROOT_PASSWORD=123123 -v "$PWD"/data:/var/lib/mysql mysql:5.6
~~~

~~~
$ ll data/
total 221192
drwxr-xr-x   9 999        999     306B Nov  4 00:23 .
drwxr-xr-x   3 krasilich  staff   102B Nov  3 19:02 ..
-rw-rw----   1 999        999      56B Nov  4 00:20 auto.cnf
-rw-rw----   1 999        999      48M Nov  4 00:23 ib_logfile0
-rw-rw----   1 999        999      48M Nov  4 00:20 ib_logfile1
-rw-rw----   1 999        999      12M Nov  4 00:23 ibdata1
drwx------  81 999        999     2.7K Nov  4 00:20 mysql
drwx------  55 999        999     1.8K Nov  4 00:20 performance_schema
drwx------   3 999        999     102B Nov  4 00:21 test_docker
~~~

And just check inside container
~~~
$ id mysql
uid=999(mysql) gid=999(mysql) groups=999(mysql)
~~~

**Comment by joonasy** [comment]
+1

**Comment by sampattuzzi** [comment]
@krasilich, Fantastic workaround. Thanks for sharing.

**Comment by pedrofaria** [comment]
@krasilich THANKS SO MUCH! this solution works very well here!

**Comment by thSoft** [comment]
@krasilich Thank you! The last command to run should be `eval "$(docker-machine env dev-nfs)"`. Otherwise, this workaround works for me, but it is very inconvenient because another VM has to be created so Docker Quickstart Terminal and Kitematic cannot be used; what's more, multiple commands have to be issued to set it up.

**Comment by bahaaldine** [comment]
+1 Works for me too !!

**Comment by oscarvs** [comment]
@krasilich +1 +1 works very well! thanks for sharing

**Comment by danawoodman** [comment]
@krasilich I owe you a beer! Worked great! :beers:

**Comment by smarques** [comment]
Thank you, @krasilich  this is very useful. Still I hope it will be fixed asap.
BTW you wrote 

     eval "$(docker-machine env nfs)"

but shouldn't it be 

   eval "$(docker-machine env dev-nfs)"

**Comment by pdonorio** [comment]
Came across a *workaround* (at least for me) recently which kinda of changed my life with Docker an OSX so i thought i'd share here, even if it doesn't solve directly the subject of the issue
It makes me work with docker differently instead, so that i don't get to mount osx host volume.

I am using the [new docker **volumes** cli option](https://docs.docker.com/engine/reference/commandline/volume_create/), available from version 1.9+.

It looks to me as a huge step forward since datacontainers:

```bash
bash-3.2$ docker --version
Docker version 1.9.1, build a34a1d5
bash-3.2$ docker volume create --name pgdata
pgdata
bash-3.2$ docker volume ls
DRIVER              VOLUME NAME
local               pgdata
```

This was my problem with postgres image mounting:

```bash
# Run postgres
bash-3.2$ docker run -d -v $(pwd)/PGDATA:/var/lib/postgres/data postgres
10ce7c1b28dee05abf3bc76493d425326ff26feb2fd6ff4d460f2368c5443d2d
# Change on the latest container permission on the data dir just mounted
bash-3.2$ docker exec $(docker ps -l -q) chown -R postgres:postgres /var/lib/postgres
# Check it...
bash-3.2$ docker exec $(docker ps -l -q) ls -l /var/lib/postgres
total 0
drwxr-xr-x 1 1000 staff 68 Nov 26 10:06 data
```
So even trying to fix the ownership permissions are unchanged...

Ok, here comes the change: instead of mounting the osx directory, i just use the volume i previously created above. And the problem is gone:

```bash
bash-3.2$ docker run -d -v pgdata:/var/lib/postgres/data postgres
a6e7c3908d8a4b0877db9b1cfd81f95466983383ee769c5363d3044e04985e43
bash-3.2$ docker exec $(docker ps -l -q) chown -R postgres:postgres /var/lib/postgres
bash-3.2$ docker exec $(docker ps -l -q) ls -l /var/lib/postgres
total 4
drwxr-xr-x 2 postgres postgres 4096 Nov 26 10:54 data
```

The volumes don't go away if you destroy the container and you can create another one and mount the same content, so persistence is preserved.

It's also easy to check the content:
```bash
bash-3.2$ docker volume inspect pgdata
[
    {
        "Name": "pgdata",
        "Driver": "local",
        "Mountpoint": "/mnt/sda1/var/lib/docker/volumes/pgdata/_data"
    }
]
bash-3.2$ docker-machine ssh dev sudo ls /mnt/sda1/var/lib/docker/volumes/pgdata/_data
```

Hope it helps someone while waiting for this issue to be solved.
Paolo

**Comment by krasilich** [comment]
Hello @pdonorio thanks for sharing.
What you are doing is simply the same if we do volume mapping to directory outside of /Users.
For example:
~~~
$docker-machine ssh dev mkdir -p /var/docker-volume/pgdata
$docker run -d -v /var/docker-volume/pgdata:/var/lib/postgresql postgres
$docker-machine ssh dev ls -la /var/docker-volume/pgdata
~~~

So our data placed to /var/docker-volume/pgdata inside boot2docker, it isn't destroyed, permissions just works and so on...
But you still can't access you data without ssh to boot2docker.

As for me it's ok to go that way until you have to edit data directly from mac os.
I use such approach for data storage containers (MySQL, Mongo, Redis...)

For application containers (PHP and Python) I use nfs approach, so I clone project to my mac, map that folder to container, edit code directly with IDE and look for results inside container.

**Comment by krasilich** [comment]
@smarques thanks! Fixed =)

As for me it will never be fixed, because the issue here is not related to boot2docker neither docker itself.
VirtualBox shared folder implementation is responsible. And first time I faced this issue with vboxsf was years and years ago... Nothing was changed since that time.

So, my suggestion to @lynxnathan or @tianon is to close this issue as won't fix

**Comment by pdonorio** [comment]
> For application containers (PHP and Python) I use nfs approach, so I clone project to my mac, map that folder to container, edit code directly with IDE and look for results inside container.

@krasilich yes, been there. I am working with python and a dev user inside the container. 

But mapping that folders as read-only inside the container is just fine, they need to be edited from my editor on osx side, and then run inside the container. So there should be no need for NFS even if you use vbox mount. Am i missing something?

By the way Virtualbox shared folders have been hunting me since first ubuntu/windows fights inside my pc...

**Comment by krasilich** [comment]
@pdonorio 
> But mapping that folders as read-only inside the container is just fine, they need to be edited from my editor on osx side, and then run inside the container. So there should be no need for NFS even if you use vbox mount. Am i missing something?

Yes, it works until your application has no need to write something to it's own folder, but usually (maybe it's me unlucky to work only with such kind of applications) they want to create folder for cache and/or session files, any kinds of imports/exports and so on.

**Comment by koryonik** [comment]
Thanks @krasilich, nice nfs workaround !

**Comment by sebwebdev** [comment]
I have an php-fpm box running on OSX. As mentioned by @saada this fixed the issue for me:
RUN usermod -u 1000 www-data
RUN usermod -G staff www-data

Thanks!

**Comment by liviubalan** [comment]
If you want learn how to mount vboxsf on Linux OS start for an HTTP server on Ubuntu Server you can take a look at <a href="http://www.liviubalan.com/mount-vboxsf-on-linux-os-start-for-an-http-server-on-ubuntu-server">Mount vboxsf on Linux OS start for an HTTP server on Ubuntu Server (video tutorial)</a>.

**Comment by rjain15** [comment]
@pdonorio I have followed your steps to create docker volumes cli approach
````
$docker volume create --name data
$docker volume ls
$docker run --name mongo -v data:/data/db -d mongo
$docker exec $(docker ps -l -q) chown -R mongodb:mongodb /data/db
$docker exec $(docker ps -l -q) ls -l /data/db
$docker volume inspect data
$docker-machine ssh default sudo ls /mnt/sda1/var/lib/docker/volumes/data/_data
````

As @krasilich mentioned, this is creating the data volume inside the boot2docker vm, I want to create them on my OSX host. 

Has anyone figured a way out?

**Comment by pdonorio** [comment]
@krasilich @rjain15 about sessions / cache you can still create other volumes or fit all inside the same volume. The important thing is to have persistence, doesn't matter where as long as you have control. And you can write stuff in this volumes if you apply the user id as i said in my previous posts.

Of course I note that if you loose your Vbox your volumes get lost too, so we may have import/export/backup [like suggested for data containers](https://docs.docker.com/engine/userguide/dockervolumes/#backup-restore-or-migrate-data-volumes):

```bash
# Export 'myvolume' content
$ docker run -v $(pwd):/backup -v myvolume:/dbdata ubuntu tar cvf /backup/backup.tar /dbdata

# Import from another Vbox
$ docker run -v $(pwd):/backup -v myvolume:/dbdata ubuntu bash -c "cd /dbdata && tar xvf /backup/backup.tar"
```

The point is that any data volume is not content you might want to edit, is just stuff you want to store somewhere. I think all possible actions can be transformed into pure docker volumes, avoiding to mount from the OSX host. I also think that volumes will give more options such as backup natively in future docker releases, as it happened for all the other stuff so far.

Finally: you might also write a script in cron to mount interested volumes and back them up. 

I know it's still more work than before, but also cleaner than having sparse directories all around your file system.

**Comment by smarques** [comment]
@krasilich solution worked wonderful for me, but today out of the blue I get:
conflicting exports for /Users, 192.168.99.100
exports:4: export option conflict for /Users

anyone had this problem? TIA

**Comment by ariault-blanchard** [comment]
What does it work with a docker-compose file ?

Actually i'm building a new docker image with Centos and the apache user inside is not allow to write...

Can somebody meet the problem ?

**Comment by ouyangpeter** [comment]
krasilich +1

**Comment by nhooey** [comment]
I get the same issue as @smarques did when he used @krasilich's solution at https://gist.github.com/mattes/4d7f435d759ca2581347

I get this error:
```bash
$ ruby <(curl -L -s https://git.io/vvvco) default
Get vboxnet ip addres ... 192.168.99.1
Update /etc/exports ...
conflicting exports for /Users, 192.168.99.100
exports:3: export option conflict for /Users
```

**Comment by nhooey** [comment]
I figured it out, I had an extra line in `/etc/exports` that conflicted. Just delete all lines in that file that have your Docker Machine IP address in them, so they don't conflict with the lines added by @krasilich's script.

@smarques

**Comment by odrzutowiec** [comment]
Hello I am using php:7-apache docker container. I had problem with php writing files to mounted directory. Solution by @sebwebdev works perfectly for me.

```
RUN usermod -u 1000 www-data
RUN usermod -G staff www-data
```

**Comment by motin** [comment]
A related workaround, showing how image maintainers can make non-root users write to locally mounted folders on osx: https://github.com/docker-library/mysql/issues/99#issuecomment-145665645

**Comment by wpp** [comment]
Big thank you to @krasilich! Also it seems like adlogix/docker-machine-nfs#17 has been fixed too.

> docker-machine-nfs test --shared-folder=/var/www --nfs-config="-alldirs -maproot=0"

**Comment by chaintan17** [comment]
You can also set share folder in VirtualBox
with auto-mount and give them full access.

**Comment by smarques** [comment]
@nhooey thank you will try that :smile:

**Comment by pdonorio** [comment]
For Mac users a recent alternative is to use [xhyve](https://github.com/mist64/xhyve) lightweight OS X virtualization solution instead of boot2docker/virtualbox.

It is very simple to install and use it with Docker with the help of the binary [dlite](https://github.com/nlf/dlite). I read and followed the instructions on [this post](https://blog.andyet.com/2016/01/25/easy-docker-on-osx/) and I was quickly up and running.

Quoting:
"DLite also mounts your entire /Users directory into the virtual machine via NFS", and (i tested it successfully) you can change permissions just fine.

**Comment by koushikckm** [comment]
Is this issue still open for Mac OS X->Virtualbox-> mysql?
I m really struggling on this

**Comment by koushikckm** [comment]
@nachimehta Were u able to overcome this issue on OS X?

**Comment by haobird** [comment]
Here is a solution。
1.运行命令：`docker-machine ls`	查看当前你使用的虚拟机
2.根据名字进入虚拟机：`docker-machine ssh default`
3.在虚拟机中编辑这个文件：`sudo vi /mnt/sda1/var/lib/boot2docker/bootlocal.sh`
4.在文件中写入如下命令：
```
#!/bin/sh

if [ -e /Users/jie ]; then
        sudo umount /Users
fi

sudo mount -t vboxsf -o uid=1001,gid=50 Users /Users
```

5.保存后，重启虚拟机：`sudo reboot`
6.docker run 新的container即可。如：`docker run -it -P --name web -v /Users/jie/DockerWork:/www ubuntu14.04:latest /bin/bash`
￼

**Comment by KIVagant** [comment]
@haobird, да, всё стало сразу понятнее, продолжай в том же духе! :)

**Comment by haobird** [comment]
@KIVagant I hope it will help someone. ^_^

**Comment by emaiax** [comment]
@mnapoli I'm still experiencing the same issue with dinghy. https://github.com/codekitchen/dinghy/issues/31

Can you tell me how have you accomplished this? :/

**Comment by mnapoli** [comment]
@emaiax I have no issues at all with dinghy :/ I just installed it following the readme and did nothing specific.

**Comment by istinspring** [comment]
holy crap i can't even imagine how such a basic things are still broken.

So it's basically impossible to use docker on OSX for development? (if mongo image can't use your local database)

**Comment by denji** [comment]
@istinspring +1

**Comment by Moghedrin** [comment]
@istinspring If it's basic, make a PR ;D Open Source ftw!

**Comment by JulienD** [comment]
@mnapoli, Issuing the same problem than @emaiax. Trying to install Mysql to a shared directory with Dinghy make Mysql crash due to a permission problem. I would be interested how you get rid of the problem. Looking at dinghy there is still an issue about this point https://github.com/codekitchen/dinghy/issues/31

**Comment by KIVagant** [comment]
@JulienD, use LXC / LXD and/or Vagrant in OSX.

**Comment by pierrerigal** [comment]
@JulienD : I had the same problem, I switched to dlite https://github.com/nlf/dlite, it just works.

**Comment by paolomainardi** [comment]
Guys to solve all the permissions problems you just have to switch to NFS
instead of vboxfs, it''s not really a matter of dlite or whatever.

Just Google: docker-machine-nfs
Il 10/mar/2016 15:04, "Pierre" <notifications@github.com> ha scritto:

**Comment by denji** [comment]
@paolomainardi Another 3rd module hack, it doesn't fix the underlying architectural problem.

**Comment by paolomainardi** [comment]
The architecture doesn't have any problem guys, the problem here is vboxfs,
just change it to NFS.
Il 10/mar/2016 16:29, "Denis Denisov" <notifications@github.com> ha scritto:

**Comment by JulienD** [comment]
@paolomainardi, any hint on how to do that ?

**Comment by paolomainardi** [comment]
Yes, you can automate using this script:
https://github.com/adlogix/docker-machine-nfs

It's a a matter of just a few commands.
Il 13/mar/2016 13:39, "Julien Dubreuil" <notifications@github.com> ha
scritto:

**Comment by binabot** [comment]
`RUN usermod -u 1000 www-data` Worked great !

**Comment by agborkowski** [comment]
@binabot is this command should run inside container ? Im using kitematic and there is not to much configuration to run command after start the container..

**Comment by binabot** [comment]
@agborkowski u can run `usermod -u 1000 www-data` inside your container OR insert `RUN usermod -u 1000 www-data` in your Dockerfile. Never used kitematic so dont know what to do with your situation ;) gl

**Comment by tanhaa** [comment]
I would really suggest mounting your shared folder as NFS as it's a better solution than individually changing your uid/gid for the containers that need access.  There are several comments in this thread that point to how to do that.

**Comment by ivicaned** [comment]
I used `RUN usermod -u 1000 www-data` into Docker file and it worked great.
I am waiting for Docker for OSX which is currently in beta, which should fix this issue with permissions: https://blog.docker.com/2016/03/docker-for-mac-windows-beta/

**Comment by SydOps** [comment]
Thanks, @ivicaned . Your fix is perfect. 

You save me the whole day that I stuck on this issue.

**Comment by SydOps** [comment]
and @ivicaned 

Regarding native docker for OSX , I already got token and installed. Yes, there is no this permission issue with same images.

**Comment by raptor235** [comment]
Hey guys could use some help here.. got everything working properly.. I'm also exposing mariadb to a mounted volume so data is persistent. 

The problem is I can set everything up... run docker-compose build, then docker-compose up, I see the db folder gets built by mariadb (those files are restricted to osx root only user) and things are running fine.

However when I run another docker-compose build command I run into issues from my php container complaining about access to the db folder which is setup from mariadb container... I'm not sure if this has something to do with /etc/exports and permissions of that folder but I can't build anything else unless I delete the db folder in the root... 

I'm not sure if any of the options for docker-machine-nfs will help with this issue...

<pre>
`#docker-machine ip default

web:
    build: .
    container_name: web
    ports:
        - "9090:80"
    volumes:
        - /Users/bartdabek/Sites/hgv/hgv_data/sites/:/code
        - ./site.conf:/etc/nginx/conf.d/site.conf
        - ./nginx.conf:/etc/nginx/nginx.conf
        - ./sites-enabled:/etc/nginx/sites-enabled
    links:
        - php
        #- mariadb
        - memcached
php:
    container_name: php
    build: .
    dockerfile: Dockerfile-php
    volumes:
        - /Users/bartdabek/Sites/hgv/hgv_data/sites/:/code
        - ./Tideways.php:/usr/local/lib/php/extensions/no-debug-non-zts-20151012/Tideways.php
    links:
        #- mariadb
        - memcached
memcached:
  container_name: memcached
  image: sameersbn/memcached:latest
  ports:
    - "11211:11211"
  restart: always
mariadb:
  container_name: mariadb
  image: mariadb
  volumes:
    - /Users/bartdabek/Sites/docker/db/:/var/lib/mysql
  environment:
    DB_ADMIN_PASS: pass
    MYSQL_ROOT_PASSWORD: root
  ports:
    - "3306:3306"
`
</pre>
![image](https://cloud.githubusercontent.com/assets/313825/15381873/9778c1fa-1d51-11e6-84e8-e3eddf86b7d0.png)

![image](https://cloud.githubusercontent.com/assets/313825/15381883/b63f4848-1d51-11e6-86f0-eb9cf4f1b7e7.png)

**Comment by raptor235** [comment]
Basically I need to change permissions back on the db folder to get things to build again 

<code>
drwxr-xr-x  11 bartdabek  staff    374 19 May 00:02 .
drwxr-xr-x  13 bartdabek  staff    442 16 May 15:36 ..
-rw-r--r--   1 bartdabek  staff    365 18 May 22:58 Dockerfile
-rw-r--r--   1 bartdabek  staff    267 18 May 22:26 Dockerfile-data
-rw-r--r--   1 bartdabek  staff   1642 18 May 23:06 Dockerfile-php
-rw-r--r--@  1 bartdabek  staff  43748 17 May 21:44 Tideways.php
drwxr-xr-x  11 999        999      374 19 May 00:02 db
-rw-r--r--   1 bartdabek  staff   1006 19 May 00:02 docker-compose.yml
-rw-r--r--   1 bartdabek  staff    743 18 May 23:09 nginx.conf
-rw-r--r--   1 bartdabek  staff    508 16 May 15:38 site.conf
drwxr-xr-x   3 bartdabek  staff    102 18 May 23:13 sites-enabled
</code>

**Comment by motin** [comment]
> A related workaround, showing how image maintainers can make non-root users write to locally mounted folders on osx: docker-library/mysql#99 (comment)

Check this out. Many seem to have this problem with mysql/mariadb and a workaround has been around for quite some time - using usermod to change the user/group ids to match the ownership of the host volume folder:

Step 1: Add the script `run.sh` somewhere in your project: [stack/localdb/run.sh](https://github.com/neam/docker-stack/blob/1a741b0fb2ce59e9c0cadf4516dc61f147e3efc2/stacks/debian-php-nginx.dna-project-base/stack/localdb/run.sh)

Step 2: Change the `command` in docker-compose.yml to run this script instead of the ordinary command. 

Example:

```
# Local database server to mimic a cloud database
localdb:
  image: mysql:5.6.27
  volumes:
    - ./stack/localdb/.db/mysql:/var/lib/mysql:rw
    - ./stack/localdb/:/stack/localdb:rw
  ports:
    - "3306"
  environment:
    MYSQL_ROOT_PASSWORD: "local-mysql-pass"
  command: "/stack/localdb/run.sh"
```

**Comment by raptor235** [comment]
Thanks @motin that worked! 

However I'm now stuck as the maridb hostname is not longer linked to any of the other containers... any idea why that default functionality is no longer working?

**Comment by raptor235** [comment]
Nevermind last comment... some sort of hickup working again! Thanks

**Comment by cpapidas** [comment]
I have a container with apache, php and mysql. The solution for me was 

> usermod -u 1000 www-data

**Comment by jolic** [comment]
@cpapidas

Could you please send a small info or a step by step info how you did this?

**Comment by cpapidas** [comment]
Of course i can

1. Run your docker image. For my case the command is
    `$ docker run -it -p 80:80 -v .../Project/PHP:/root/project myImageName`
    
    We are in docker VM now

2. Run  `$ usermod -u 1000 www-data `

3. Then if you to see the files' permissions, will be similar with image bellow

![screen shot 2016-06-20 at 12 16 42 am](https://cloud.githubusercontent.com/assets/5034374/16179893/a4006358-367c-11e6-8046-bc949d4b3628.png)

**Comment by ababushkin** [comment]
Has anyone worked out a simple - StackOverflow style - solution to this problem yet?

I've tried all the conventional solutions, such as:

1. Creating and configuring the permissions script (as per @motin's) comment
2. Manually running usermod -u 1000 mysql in a custom Dockerfile (that inherits from this one)

The only 'feasible' solution I see is a custom project created by @dgraziotin, which deviates from the main MySQL / MariaDB docker images (https://github.com/dgraziotin/osx-docker-mysql). This hardly seems like an optimal solution, especially if Docker is to get even more rapid adoption throughout the community.

**Comment by ernestom** [comment]
@ababushkin using docker-machine-nfs worked for me.

**Comment by yosifkit** [comment]
With https://github.com/docker-library/mysql/pull/161 you should be able to run mysql as the owner of the directory in question:
```console
docker run -d -e MYSQL_ROOT_PASSWORD=foobar1234 --user 1000:50 -v /Users/my-user/mysql-data/:/var/lib/mysql mysql:5.7
```
This will fix the permissions problem, but I cannot guarantee that mysql will always run on a VirtualBox Shared Folder.  MongoDB for example cannot run on the shared folder since the file system does not support everything it needs.

**Comment by motin** [comment]
@ababushkin I have read reports from users of [native Docker for Mac](https://www.docker.com/products/docker#/mac) that they no longer need any workaround. I created the permissions script to make it work on Docker Toolbox.

**Comment by dgraziotin** [comment]
@ababushkin I confirm that with Docker v1.12.0 for Mac, there is no need to use my ugly workaround anymore :-) @motin

**Comment by ababushkin** [comment]
@dgraziotin @motin Ouch, looks like I'm still running Docker v1.11.0.

I'll upgrade and give it a whirl.

@yosifkit thanks for the tip!

**Comment by ababushkin** [comment]
@ernestom did you notice any performance improvement for disk sync when using that solution? Symfony runs really poorly for me at the moment. I have the same issue when using Vagrant and VirtualBox shared volumes and worked around it by using an NFS mount.

**Comment by ernestom** [comment]
@ababushkin I didn't notice any significant impact in performance with NFS for Docker, and I've been using it for years on my vagrant/vbox projects without issues.

**Comment by krasilich** [comment]
@ababushkin I have been using Docker for Mac for a couple of months now, and unfortunately performance issues still there for me. My case is - project on local, mounted inside container, running inside container. For example some project on Symfony, I run some heavy with high I/O app/console command and waiting tens of minutes or even hours to complete instead of up to 10 minutes on pure local.

**Comment by ababushkin** [comment]
@ernestom @krasilich Thats a bummer.

@ernestom do you have a boilerplate docker-compose file that's using the NFS solution by any chance?

**Comment by dend** [comment]
So just to follow-up here, the issue seems to be gone on the Mac if you install the native Docker beta (use [beta channel here](https://docs.docker.com/docker-for-mac/)).

**Comment by ababushkin** [comment]
@dend Yup thats correct, permission issues are fixed with Docker for Mac. You don't need to install the beta version, the stable release fixes it as well. 

Unfortunately performance issues have not been fixed. At this stage I've been getting around performance issues by doing smarter builds of my images (e.g. mounting volumes that don't need lots of read/write operations by the app)

**Comment by vschoener** [comment]
Just trying back to use Docke for mac with a web project, the ownership issue is still here..
Files are mounted in root:root and even if I change the ownership to www-data with a entrypoint, the new one will be created as root user.

Any idea ? I tried to use USER www-data before starting the apache process but as you know, www-data doesn't have the privilege to start the apache service.. Feel stuck :(

**Comment by TyIsI** [comment]
Ran into this issue as well.

@vmaatta was right in his breakdown, and I'd like to add that the "issue" is the -r option in groupadd/useradd versus adding users without that specific option. The -r option creates system users, which by default (set in /etc/adduser.conf) "starts" with UID/GID 999. (Last in range.)

```
root@9b3da358b593:/# egrep SYSTEM /etc/adduser.conf
# FIRST_SYSTEM_[GU]ID to LAST_SYSTEM_[GU]ID inclusive is the range for UIDs
FIRST_SYSTEM_UID=100
LAST_SYSTEM_UID=999
FIRST_SYSTEM_GID=100
LAST_SYSTEM_GID=999
```

For regular users, these would be added starting UID/GID 1000 (matching the UID for the boot2docker docker user), which is fine for a single user. This also has the implication that if another user were to be added in a container, that a user with UID 1001 wouldn't be able to access files through vboxsf.

Right now, I don't know how this could be solved easily, but I'm going to look into this.

Recap: Images create system users that don't match the UID of the _docker_ user in boot2docker.

**Comment by TyIsI** [comment]
Example work-around for rabbitmq:

Dockerfile:
```
FROM rabbitmq:3-management

RUN usermod -u 1000 rabbitmq
RUN groupmod -g 1000 rabbitmq
```

```docker build -t rabbitmq-test .```

```docker run -d --hostname my-rabbit --name some-rabbit -p 8080:15672  -v `pwd`/data/rabbitmq/:/var/lib/rabbitmq rabbitmq-test```

**Comment by yosifkit** [comment]
@TyIsI, many of the images provided by Docker Official Images (like `rabbitmq`) were modified to allow running as a different user so that you would not need to create or modify the user in the image. See https://github.com/docker-library/rabbitmq/pull/60 and the other PRs linked from there.  What this means is that in most instances when using boot2docker on a Mac, you can do something like the following:

```console
$ docker run -d -v /Users/myuser/rabbitdir/:/var/lib/rabbitmq/ --user 1000:50 rabbitmq:3-managment
```

Some notable exceptions that don't work with the VirtualBox shared folder are MongoDB (https://github.com/docker-library/mongo/issues/74#issuecomment-175855893) and MariaDB 10.1 (https://github.com/docker-library/percona/issues/42#issuecomment-297865389).
